src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,193 @@
+        last.next = newNode;
+        newNode.next = null;
+        last = newNode;
+    }
+
     @Override
     public void add(T value) {
-
+        if (size == 0) {
+            addFirst(value);
+        } else {
+            addBack(value);
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (size < index || index < 0) {
+            throw new IndexOutOfBoundsException();
+        }
+        if (index == 0) {
+            if (size == 0) {
+                addFirst(value);
+            } else {
+                addFront(value);
+            }
+        }
+        if (index > 0) {
+            if (index == size) {
+                addBack(value);
+            }
+            newNode = new Node(value);
+            nodeByIndex(index - 1).next = newNode;
+            newNode.next = nodeByIndex(index);
+            newNode.prev = nodeByIndex(index - 1);
+            nodeByIndex(index).prev = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+            size++;
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (size <= index || index < 0) {
+            throw new IndexOutOfBoundsException();
+        }
+        return nodeByIndex(index).item;
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (size <= index || index < 0) {
+            throw new IndexOutOfBoundsException();
+        }
+        nodeByIndex(index).item = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> nodeToBeRemoved = null;
+        if (size <= index || index < 0) {
+            throw new IndexOutOfBoundsException();
+        }
+        if (size == 1) {
+            first = last = null;
+            size--;
+            return null;
+        }
+        if (index == 0) {
+            nodeToBeRemoved = first;
+            first.next.prev = null;
+            first = null;
+            first = nodeToBeRemoved.next;
+        }
+        if (index > 0) {
+            if (index == size - 1) {
+                nodeToBeRemoved = last;
+                last.prev.next = null;
+                last = null;
+                last = nodeToBeRemoved.prev;
+            } else {
+                nodeToBeRemoved = nodeByIndex(index);
+                nodeByIndex(index - 1).next = nodeByIndex(index + 1);
+                nodeByIndex(index + 1).prev = nodeByIndex(index - 1);
+                nodeByIndex(index).prev = null;
+                nodeByIndex(index).next = null;
+            }
+        }
+        size--;
+        if (size == 0) {
+            throw new IndexOutOfBoundsException();
+        }
+        return nodeToBeRemoved.item;
     }
 
     @Override
     public T remove(T t) {
-        return null;
+        newNode = first;
+        int count = 0;
```

Comment:
can we remove `count` and use `i`?

Reply:
to return null we can use i, but to use remove() I will need some count anyway, because i is only inside for()

Reply:
we can write:
int i = 0;
for(i; i< size; i++){
...
and then we can use `i` and remove `count`

Reply:
not works. wants to define i inside scope